### DISCUSSION TITLE
table.metadata.create_all() and support for SQL-Server ColumnStore?

### DISCUSSION THREAD

**Discussion by MarcusJellinghaus** [sqlalchemy/sqlalchemy#7333] [Usage Questions] [answered]
Hello,

I have this code to create a new table based on the structure of an existing table.
```
source_engine = driver_connection.create_sqlserver_engine(source_server, source_database)
engine_new = driver_connection.create_sqlserver_engine(new_server, new_database)

import sqlalchemy

smeta = sqlalchemy.MetaData(bind=source_engine)
table = sqlalchemy.Table(source_table_name, smeta, schema=source_schema_name, autoload=True)
table.metadata.create_all(engine_new)
```

I'm using the latest SQL Server (2019) and the table is based on [columnstore](https://docs.microsoft.com/en-us/sql/relational-databases/indexes/columnstore-indexes-overview?view=sql-server-ver15),

The creation of the table does not work well. The creation of the columnstore index fails.

Is this expected to be supported? Did I overlook something? Or shall I prepare python and SQL scripts to report this as an issue?

**Comment by zzzeek** [comment]
hey there -

> The creation of the columnstore index fails.

this is vague. can you be more specific ?

> Is this expected to be supported?

The page you linked does not refer to any SQL syntax for this "columnstore" feature so I dont know what it looks like, however  SQLAlchemy does not currently have any syntax that can "create" a "columnstore" index if there is special SQL syntax required.

However, your example as written doesn't illustrate any kind of "columnstore" index being created, it is reflecting a Table object from some particular database, and the most that would happen as far as indexes there is that you'd get back some Index() objects that look like normal indexes, there would be no state within that Python object that suggests "columnstore", so other than "my new table doent have a columnstore index on it", which would be expected, I am not aware of what other kind of "failure" would be happening (such as a stack trace).

**Reply by MarcusJellinghaus** [reply]
Here is some code to create a table in SQL-Server with a columnstore index:

```
CREATE TABLE [myTable](
       [myString] [nvarchar](36) NULL,
       [myInt] [bigint] NULL,
       [myTable_ID] [bigint] IDENTITY(1,1) NOT NULL
)

CREATE CLUSTERED COLUMNSTORE INDEX [PK_CLUSTERED_myTable] ON [myTable]
```

Here is the [stacktrace.txt](https://github.com/sqlalchemy/sqlalchemy/files/7553160/table__metadata__create_all.stacktrace.txt) as well as the specific [create columnstore index documentation](https://docs.microsoft.com/en-us/sql/t-sql/statements/create-columnstore-index-transact-sql?view=sql-server-ver15).

In particular, sqlalchemy creates an SQL statement: `[SQL: CREATE INDEX [PK_CLUSTERED_myTable] ON dbo.[myTable] ()]`
The keywords `CLUSTERED COLUMNSTORE` are missing, the brackets `()` are incorrect.

Would you think this is a bug or a missing feature?

**Reply by CaselIT** [reply]
It seems there is both a but (the `()`) and a missing feature, the clustered columnstore.

Can you indicate what is reflected for that index by using [`get_indexes`](https://docs.sqlalchemy.org/en/14/core/reflection.html?highlight=get_index#sqlalchemy.engine.reflection.Inspector.get_indexes)?

If the index type is reflected correctly, it may just be a missing compiler feature. Otherwise we also need to update the reflection logic

**Reply by zzzeek** [reply]
it looks like it is reflecting the index but not getting any columns back, that's all.  

first order of business would be that this index has to be skipped upon reflection, like other indexes for which we do not have a Python construct such as PG expression based indexes.

**Reply by MarcusJellinghaus** [reply]
Thank you for the clarification.

Here is a small SQL query that I use to detect the index type - maybe it is useful for you:
```
 SELECT       
        --SCHEMA_NAME(sys.tables.[schema_id]) as schema_name,
        --sys.tables.name as table_name,
      --sys.indexes.name as index_name,
        --sys.indexes.is_primary_key as IsPrimaryKey,
      --sys.indexes.type as type,
      CASE
         WHEN sys.indexes.type = 0 THEN 'Heap'
         WHEN sys.indexes.type = 1 THEN 'Clustered'
         WHEN sys.indexes.type = 2 THEN IIF(is_unique_constraint = 0, 'Nonclustered index', 'Unique constraint')
         WHEN sys.indexes.type = 3 THEN 'XML'
         WHEN sys.indexes.type = 4 THEN 'Spatial'
         WHEN sys.indexes.type = 5 THEN 'Clustered columnstore'
         WHEN sys.indexes.type = 6 THEN 'Nonclustered hash index'
         ELSE 'unknown type'
      END as type_info
        --, sys.indexes.is_unique_constraint
    FROM sys.indexes -- AS I
    INNER JOIN sys.tables -- AS T
        ON sys.indexes.[object_id] = sys.tables.[object_id]
    WHERE
        sys.indexes.type_desc <> N'HEAP'
      AND SCHEMA_NAME(sys.tables.[schema_id]) = '{schema}'
      AND sys.tables.name = '{table_name}'
      AND sys.indexes.name = '{index_name}'
```

**Reply by CaselIT** [reply]
if you want to look at what we currently use you find it here:

https://github.com/sqlalchemy/sqlalchemy/blob/6206f0ff74e95c9339dc0f0e26caab55e9bcda45/lib/sqlalchemy/dialects/mssql/base.py#L3005-L3073

**Comment by CaselIT** [comment, marked as answer]
I think an issue would be better for this

**Reply by CaselIT** [reply]
let's continue in the issue #7340
